dvc/pkg.py

```diff
@@ -0,0 +1,160 @@
+            cache_config.set_dir(cache_dir, level=Config.LEVEL_LOCAL)
+
+    def uninstall(self):
+        if not self.installed:
+            return
+
+        shutil.rmtree(self.path)
+
+    def update(self):
+        self.repo.scm.fetch(self.version)
+
+    def dumpd(self):
+        ret = {self.PARAM_NAME: self.name}
+
+        if self.url:
+            ret[self.PARAM_URL] = self.url
+
+        if self.version:
+            ret[self.PARAM_VERSION] = self.version
+
+        return ret
+
+
+class PkgManager(object):
+    PKG_DIR = "pkg"
+
+    def __init__(self, repo):
+        self.repo = repo
+        self.pkg_dir = os.path.join(repo.dvc_dir, self.PKG_DIR)
+        self.cache_dir = repo.cache.local.cache_dir
+
+    def install(self, url, **kwargs):
+        pkg = Pkg(self.pkg_dir, url=url, **kwargs)
+        pkg.install(cache_dir=self.cache_dir)
+
+    def uninstall(self, name):
+        pkg = Pkg(self.pkg_dir, name=name)
+        pkg.uninstall()
+
+    def get_pkg(self, **kwargs):
+        return Pkg(self.pkg_dir, **kwargs)
+
+    def imp(self, name, src, out=None, version=None):
+        scheme = urlparse(name).scheme
+
+        if os.path.exists(name) or scheme:
+            pkg = Pkg(self.pkg_dir, url=name)
+            pkg.install(cache_dir=self.cache_dir)
+        else:
+            pkg = Pkg(self.pkg_dir, name=name)
+
+        info = {Pkg.PARAM_NAME: pkg.name}
+        if version:
+            info[Pkg.PARAM_VERSION] = version
+
+        self.repo.imp(src, out, pkg=info)
+
+    @classmethod
+    def get(cls, url, src, out=None, version=None):
+        if not out:
+            out = os.path.basename(src)
+
+        # Creating a directory right beside the output to make sure that they
+        # are on the same filesystem, so we could take the advantage of
+        # reflink and/or hardlink.
+        dpath = os.path.dirname(os.path.abspath(out))
+        with TemporaryDirectory(dir=dpath, prefix=".") as tmp_dir:
+            pkg = Pkg(tmp_dir, url=url, version=version)
+            pkg.install()
+            # Try any links possible to avoid data duplication.
+            #
+            # Not using symlink, because we need to remove cache after we are
+            # done, and to make that work we would have to copy data over
+            # anyway before removing the cache, so we might just copy it
+            # right away.
+            #
+            # Also, we can't use theoretical "move" link type here, because
+            # the same cache file might be used a few times in a directory.
+            pkg.repo.config.set(
+                Config.SECTION_CACHE,
+                Config.SECTION_CACHE_TYPE,
+                "reflink,hardlink,copy",
+            )
+            src = os.path.join(pkg.path, urlparse(src).path.lstrip("/"))
+            output, = pkg.repo.find_outs_by_path(src)
+            pkg.repo.fetch(output.stage.path)
+            output.path_info = PathInfo(os.path.abspath(out))
+            with output.repo.state:
```

Comment:
btw, does it make sense to disable lock and sqllite for this case? just to make sure that this easy scenario works everywhere.

Reply:
You mean nfs and friends, right? Good point, I'll take a look if we could disable state for that purpose.

Reply:
This is not trivial, created https://github.com/iterative/dvc/issues/2135 for it. Might be easier to just support nfs after all.